vllm/utils.py

```diff
@@ -510,6 +512,187 @@ def random_uuid() -> str:
     return str(uuid.uuid4().hex)
 
 
+class AsyncMicrobatchTokenizer:
+    """Asynchronous tokenizer with micro-batching.
+
+    A single background task pulls pending encode / decode requests from
+    an ``asyncio.Queue`` and groups them into micro-batches.  The actual
+    blocking tokenizer call is executed inside a ``ThreadPoolExecutor`` 
+    with a single worker, so the event-loop never blocks.
+    """
+
+    def __init__(
+        self,
+        tokenizer,
+        max_batch_size: int = 32,
+        batch_wait_timeout_s: float = 0.002,
+    ) -> None:
+        self.tokenizer = tokenizer
+        self.max_batch_size = max_batch_size
+        self.batch_wait_timeout_s = batch_wait_timeout_s
+
+        self._loop = asyncio.get_running_loop()
+        self._queues: dict[tuple, asyncio.Queue[tuple[str, object, dict,
+                                                      asyncio.Future]]] = {}
+        self._batcher_tasks: dict[tuple, asyncio.Task] = {}
+
+        # Single worker that owns the blocking tokenizer.
+        self._executor = ThreadPoolExecutor(max_workers=1)
+
+    # ------------------------------------------------------------------
+    # Public async API
+    # ------------------------------------------------------------------
+    async def __call__(self, prompt, **kwargs):
+        fut: asyncio.Future = self._loop.create_future()
+        key = self._queue_key("encode", kwargs)
+        queue = self._ensure_queue_and_batcher(self._loop, key)
+        await queue.put(("encode", prompt, kwargs, fut))
+        return await fut
+
+    async def decode(self, token_ids, **kwargs):
+        fut: asyncio.Future = self._loop.create_future()
+        key = self._queue_key("decode", kwargs)
+        queue = self._ensure_queue_and_batcher(self._loop, key)
+        await queue.put(("decode", token_ids, kwargs, fut))
+        return await fut
+
+    # ------------------------------------------------------------------
+    # Internal helpers
+    # ------------------------------------------------------------------
+    def _ensure_queue_and_batcher(self, loop: asyncio.AbstractEventLoop,
+                                  key: tuple):
+        """Return the queue for key, creating queue 
+        and batcher task if needed."""
+        if key not in self._queues:
+            self._queues[key] = asyncio.Queue()
+
+        if key not in self._batcher_tasks or self._batcher_tasks[key].done():
+            self._batcher_tasks[key] = loop.create_task(
+                self._batch_loop(self._queues[key]))
+
+        return self._queues[key]
```

Comment:
Could we make the batcher tasks just be a list? And just have the task wait on the queue indefinitely? I don't think the task needs to exit

```suggestion
    def _get_queue(self, key: tuple):
        """Return the queue for key, creating queue 
        and batcher task if needed."""
        queue = self._queues.get(key)
        if queue is None:
            self._queues[key] = queue = asyncio.Queue()
            self._batcher_tasks.append(
                loop.create_task(self._batch_loop(queue)))

        return queue
```

Reply:
updated